File: kotlin-remix-run-router/karakum/plugins/convertUtilityTypes.js
import ts from "typescript"

export default function (node, context, render) {
    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "Partial"
            || node.typeName.text === "Readonly"
            || node.typeName.text === "Exclude"
            || node.typeName.text === "Pick"
            || node.typeName.text === "Omit"
            // custom
            || node.typeName.text === "RequireOne"
        )
    ) {
        return render(node.typeArguments[0])
    }
    return null
}

File: kotlin-tanstack-virtual-core/karakum/src/jsMain/kotlin/wrappersgenerator/tanstack/virtual/core/plugins/convertUtilityTypes.kt
package wrappersgenerator.tanstack.virtual.core.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.asArray
import typescript.isIdentifier
import typescript.isTypeReferenceNode

// `Required<T>`
val convertUtilityTypes = createPlugin { node, _, render ->
    nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Required")

        val typeArgument = ensureNotNull(node.typeArguments?.asArray()?.singleOrNull())

        render(typeArgument)
    }
}
